src/term_translator.cpp

```diff
@@ -74,82 +113,162 @@ Term TermTranslator::transfer_term(const Term & term)
+          cache[t] = value_from_smt2(t->to_string(), t->get_sort());
         }
       }
-      else if (cached_children.size())
-      {
-        cache[t] = solver->make_term(t->get_op(), cached_children);
-      }
-      else if (t->is_symbol())
-      {
-        // already created symbol and added to cache
-        continue;
-      }
       else
       {
-        throw SmtException("Can't transfer term: " + t->to_string());
+        assert(!t->is_symbol());
+        assert(!t->is_param());
+        assert(!t->is_value());
+        assert(!t->get_op().is_null());
+
+        cached_children.clear();
+        for (auto c : t)
+        {
+          cached_children.push_back(cache.at(c));
+        }
+        assert(cached_children.size());
+
+        Op op = t->get_op();
+        if (!check_sortedness(op, cached_children))
+        {
+          /* NOTE: interesting behavior here
+             if transferring between two solvers that alias sorts
+             e.g. two different instances of BTOR
+             the sorted-ness check will still fail for something like
+             Ite(BV{1}, BV{8}, BV{8})
+             so we'll reach this point and cast
+             but the cast won't actually do anything for BTOR
+             in other words, check_sortedness is not guaranteed
+             to hold after casting */
+          cache[t] = cast_op(op, cached_children);
+        }
+        else
+        {
+          cache[t] = solver->make_term(t->get_op(), cached_children);
+        }
       }
     }
   }
 
-  return cache[term];
+  assert(cache.find(term) != cache.end());
+  // make sure the sort is as-expected and cast if not
+  // for dealing with solvers that alias sorts
+
+  return cache.at(term);
+}
+
+Term TermTranslator::transfer_term(const Term & term, const SortKind sk)
+{
+  Term transferred_term = transfer_term(term);
+  Sort transferred_sort = transferred_term->get_sort();
+  SortKind transferred_sk = transferred_sort->get_sort_kind();
+  if (transferred_sk == sk)
+  {
+    return transferred_term;
+  }
+
+  // only handles Bool <-> BV1, and Int <-> Real
+  // otherwise throws an exception
+
+  // expect this to be the most common case
+  if (transferred_sk == BV && transferred_sort->get_width() == 1 && sk == BOOL)
+  {
+    Sort sort = solver->make_sort(BOOL);
+    return cast_term(transferred_term, sort);
+  }
+  else if (transferred_sk == BOOL && sk == BV)
+  {
+    Sort sort = solver->make_sort(BV, 1);
+    return cast_term(transferred_term, sort);
+  }
+  else if (transferred_sk == INT && sk == REAL)
+  {
+    Sort sort = solver->make_sort(REAL);
+    return cast_term(transferred_term, sort);
+  }
+  else if (transferred_sk == REAL && sk == INT)
+  {
+    Sort sort = solver->make_sort(INT);
+    return cast_term(transferred_term, sort);
+  }
+  else
+  {
+    string msg("Cannot cast ");
+    msg += transferred_term->to_string() + " to " + smt::to_string(sk);
+    throw IncorrectUsageException(msg);
+  }
```

Comment:
what happens if the sort of transferred_sk and sk are the same?

Reply:
It will just return the term unaltered: https://github.com/makaimann/smt-switch/blob/5bd71ccdbe37180d79bfdd8598dae1ad2821aa90/src/term_translator.cpp#L230